app/controllers/supplementary/ProcedureCodesPageController.scala

```diff
@@ -39,29 +41,176 @@ class ProcedureCodesPageController @Inject()(
+    userInput: ProcedureCode,
+    cachedData: ProcedureCodesData
+  )(implicit request: Request[_], hc: HeaderCarrier): Future[Result] =
+    (userInput, cachedData.additionalProcedureCodes) match {
+      case (procedureCode, Seq()) =>
+        procedureCode match {
+          case ProcedureCode(Some(procedureCode), Some(additionalCode)) =>
+            val procedureCodes = ProcedureCodesData(Some(procedureCode), Seq(additionalCode))
+
+            customsCacheService.cache[ProcedureCodesData](cacheId, formId, procedureCodes).map { _ =>
+              Redirect(controllers.supplementary.routes.SupervisingCustomsOfficeController.displayForm())
+            }
+
+          case ProcedureCode(procedureCode, additionalCode) =>
+            val procedureCodeError = procedureCode.fold(
+              Seq(("procedureCode", "supplementary.procedureCodes.procedureCode.error.empty"))
+            )(_ => Seq[(String, String)]())
+
+            val additionalCodeError = additionalCode.fold(
+              Seq(("additionalProcedureCode", "supplementary.procedureCodes.additionalProcedureCode.mandatory.error"))
+            )(_ => Seq[(String, String)]())
+
+            errorHandler(procedureCodeError ++ additionalCodeError, userInput, cachedData.additionalProcedureCodes)
+        }
+
+      case (procedureCode, seq) =>
+        procedureCode match {
+          case ProcedureCode(None, _) if !cachedData.procedureCode.isDefined =>
+            errorHandler(
+              Seq(("procedureCode", "supplementary.procedureCodes.procedureCode.error.empty")),
+              userInput,
+              cachedData.additionalProcedureCodes
+            )
+
+          case ProcedureCode(_, Some(_)) if seq.length >= 99 =>
+            errorHandler(
+              Seq(("", "supplementary.procedureCodes.additionalProcedureCode.maximumAmount.error")),
+              userInput,
+              cachedData.additionalProcedureCodes
+            )
+
+          case ProcedureCode(_, Some(value)) if seq.contains(value) =>
+            errorHandler(
+              Seq(("additionalProcedureCode", "supplementary.procedureCodes.additionalProcedureCode.duplication")),
+              userInput,
+              cachedData.additionalProcedureCodes
+            )
+
+          case ProcedureCode(Some(procedureCode), additionalCode) =>
+            val updatedCache = ProcedureCodesData(
+              Some(procedureCode),
+              cachedData.additionalProcedureCodes ++ additionalCode.fold(Seq[String]())(Seq(_))
+            )
+
+            customsCacheService.cache[ProcedureCodesData](cacheId, formId, updatedCache).map { _ =>
+              Redirect(controllers.supplementary.routes.SupervisingCustomsOfficeController.displayForm())
+            }
+        }
+    }
+  //scalastyle:on methodLength
+
+  private def retrieveProcedureCode(action: String): String = action.dropWhile(_ != ':').drop(1)
+
+  private def errorPage()(implicit request: Request[_]): Future[Result] =
```

Comment:
This and the next methods names are slightly misleading. Please, add a verb to it, like 'display'.